src/librustc_mir/monomorphize/collector.rs

```diff
@@ -434,7 +434,7 @@ fn check_recursion_limit<'tcx>(
     // Code that needs to instantiate the same function recursively
     // more than the recursion limit is assumed to be causing an
     // infinite expansion.
-    if recursion_depth > *tcx.sess.recursion_limit.get() {
+    if adjusted_recursion_depth > *tcx.sess.recursion_limit.get() {
         let error = format!("reached the recursion limit while instantiating `{}`", instance);
         if let Some(hir_id) = tcx.hir().as_local_hir_id(def_id) {
             tcx.sess.span_fatal(tcx.hir().span(hir_id), &error);
```

Comment:
I feel like we should just search for `hir().span` and replace as many of them as possible...
(this should be `tcx.def_span(def_id)` and not check for `def_id` being local, but always use `span_fatal`).